src/subclass/widget.rs

```diff
@@ -0,0 +1,249 @@
+use gtk_sys;
+
+use glib::translate::*;
+
+use glib::subclass::prelude::*;
+
+use crate::Orientation;
+use Widget;
+use WidgetClass;
+
+pub trait WidgetImpl:
+    WidgetImplExt + ObjectImpl + 'static
+{
+    fn adjust_baseline_allocation(&self, widget: &Widget, baseline: &mut i32) {
+        self.parent_adjust_baseline_allocation(widget, baseline)
+    }
+
+    fn adjust_baseline_request(&self, widget: &Widget, minimum_baseline: &mut i32, natural_baseline: &mut i32) {
+        self.parent_adjust_baseline_request(widget, minimum_baseline, natural_baseline)
+    }
+
+    fn adjust_size_allocation(&self, widget: &Widget, orientation: Orientation, minimum_size: &mut i32, natural_size: &mut i32, allocated_pos: &mut i32, allocated_size: &mut i32) {
+        self.parent_adjust_size_allocation(widget, orientation, minimum_size, natural_size, allocated_pos, allocated_size)
+    }
+
+    fn adjust_size_request(&self, widget: &Widget, orientation: Orientation, minimum_size: &mut i32, natural_size: &mut i32) {
+        self.parent_adjust_size_request(widget, orientation, minimum_size, natural_size)
+    }
+
+    fn button_press_event(&self, widget: &Widget, event: &mut gdk::EventButton) -> bool {
```

Comment:
Should this be mutable? Is there anything meaningful one can modify in the event and is that allowed?

Reply:
I decided to make it mutable in case anyone wants to use any `&mut self` functions of `gdk::EventButton`.

Reply:
The question is if that's valid / allowed by the GTK API. My gut feeling would say no :)

Reply:
👍🏾

Reply:
According to `rustc` I think it needs to be mutable, since the vfunc doesn't take an owned `gdk::EventButton` but a mutable pointer:

```
error[E0596]: cannot borrow `*event` as mutable, as it is behind a `&` reference
   --> src/subclass/widget.rs:139:48
    |
131 |     fn parent_button_release_event(&self, widget: &Widget, event: &gdk::EventButton) -> Inhibit {
    |                                                                   ----------------- help: consider changing this to be a mutable reference: `&mut gdk::EventButton`
...
139 |             Inhibit(f(widget.to_glib_none().0, event.to_glib_none_mut().0) != 0)
    |                                                ^^^^^ `event` is a `&` reference, so the data it refers to cannot be borrowed as mutable

error: aborting due to previous error
```

Reply:
`glib::translate::mut_override(event.to_glib_none().0)` or a simple cast :)